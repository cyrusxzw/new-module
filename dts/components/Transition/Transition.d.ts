import { ReactElement } from 'react';
import type { TransitionProps } from './Transition.types';
declare const Transition: ({
  children,
  isActive,
  isActiveOnMount,
  shouldMountOnEnter,
  shouldUnmountOnExit,
  type,
}: TransitionProps) => ReactElement | null;
export { Transition };
