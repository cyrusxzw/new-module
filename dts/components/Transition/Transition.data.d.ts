declare const data: {
  fade: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  shiftInDown: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  shiftInLeft: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  shiftInUp: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  slideDown: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  slideRight: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  slowFade: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
  zoom: {
    classNames: {
      enter: any;
      enterActive: any;
      enterDone: any;
      exit: any;
      exitActive: any;
      exitDone: any;
    };
    timeout: number;
  };
};
export { data };
