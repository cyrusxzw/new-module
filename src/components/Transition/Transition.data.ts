import transitions from './Transition.module.css';

const fade = {
  classNames: {
    enter: transitions.fadeEnter,
    enterActive: transitions.fadeEnterActive,
    enterDone: transitions.fadeEnterDone,
    exit: transitions.fadeExit,
    exitActive: transitions.fadeExitActive,
    exitDone: transitions.fadeExitDone,
  },
  timeout: 300,
};

const slowFade = {
  classNames: {
    enter: transitions.slowFadeEnter,
    enterActive: transitions.slowFadeEnterActive,
    enterDone: transitions.slowFadeEnterDone,
    exit: transitions.slowFadeExit,
    exitActive: transitions.slowFadeExitActive,
    exitDone: transitions.slowFadeExitDone,
  },
  timeout: 1500,
};

const shiftInLeft = {
  classNames: {
    enter: transitions.shiftInLeftEnter,
    enterActive: transitions.shiftInLeftEnterActive,
    enterDone: transitions.shiftInLeftEnterDone,
    exit: transitions.shiftInLeftExit,
    exitActive: transitions.shiftInLeftExitActive,
    exitDone: transitions.shiftInLeftExitDone,
  },
  timeout: 1500,
};

const shiftInUp = {
  classNames: {
    enter: transitions.shiftInUpEnter,
    enterActive: transitions.shiftInUpEnterActive,
    enterDone: transitions.shiftInUpEnterDone,
    exit: transitions.shiftInUpExit,
    exitActive: transitions.shiftInUpExitActive,
    exitDone: transitions.shiftInUpExitDone,
  },
  timeout: 1500,
};

const shiftInDown = {
  classNames: {
    enter: transitions.shiftInDownEnter,
    enterActive: transitions.shiftInDownEnterActive,
    enterDone: transitions.shiftInDownEnterDone,
    exit: transitions.shiftInDownExit,
    exitActive: transitions.shiftInDownExitActive,
    exitDone: transitions.shiftInDownExitDone,
  },
  timeout: 1500,
};

const slideDown = {
  classNames: {
    enter: transitions.slideDownEnter,
    enterActive: transitions.slideDownEnterActive,
    enterDone: transitions.slideDownEnterDone,
    exit: transitions.slideDownExit,
    exitActive: transitions.slideDownExitActive,
    exitDone: transitions.slideDownExitDone,
  },
  timeout: 800,
};

const slideRight = {
  classNames: {
    enter: transitions.slideRightEnter,
    enterActive: transitions.slideRightEnterActive,
    enterDone: transitions.slideRightEnterDone,
    exit: transitions.slideRightExit,
    exitActive: transitions.slideRightExitActive,
    exitDone: transitions.slideRightExitDone,
  },
  timeout: 800,
};

const slideLeft = {
  classNames: {
    enter: transitions.slideLeftEnter,
    enterActive: transitions.slideLeftEnterActive,
    enterDone: transitions.slideLeftEnterDone,
    exit: transitions.slideLeftExit,
    exitActive: transitions.slideLeftExitActive,
    exitDone: transitions.slideLeftExitDone,
  },
  timeout: 400,
};

const zoom = {
  classNames: {
    enter: transitions.zoomEnter,
    enterActive: transitions.zoomEnterActive,
    enterDone: transitions.zoomEnterDone,
    exit: transitions.zoomExit,
    exitActive: transitions.zoomExitActive,
    exitDone: transitions.zoomExitDone,
  },
  timeout: 500,
};

const data = {
  fade,
  shiftInDown,
  shiftInLeft,
  shiftInUp,
  slideDown,
  slideLeft,
  slideRight,
  slowFade,
  zoom,
};

export { data };
