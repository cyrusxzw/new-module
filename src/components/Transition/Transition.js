import React, { isValidElement } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import get from 'lodash/get';
import { CSSTransition } from 'react-transition-group';
import { useHasMounted } from '~/customHooks';
import { data } from './Transition.data';

const Transition = ({
  children,
  hasCSSTransitionMountOnEnter,
  hasCSSTransitionUnmountOnExit,
  isActive,
  isActiveOnMount,
  type,
}) => {
  const hasMounted = useHasMounted();

  if (!isValidElement(children)) {
    return null;
  }

  const isVisible = isActiveOnMount ? !!hasMounted : !!isActive;

  return (
    <CSSTransition
      classNames={get(data[type], 'classNames', '')}
      in={isVisible}
      mountOnEnter={!!hasCSSTransitionMountOnEnter}
      timeout={get(data[type], 'timeout', 300)}
      unmountOnExit={!!hasCSSTransitionUnmountOnExit}
    >
      {React.cloneElement(children, {
        className: cx(
          get(children, 'props.className', ''),
          isVisible ? '' : get(data[type], 'classNames.enter', ''),
        ),
      })}
    </CSSTransition>
  );
};

Transition.propTypes = {
  children: PropTypes.element.isRequired,
  hasCSSTransitionMountOnEnter: PropTypes.bool,
  hasCSSTransitionUnmountOnExit: PropTypes.bool,
  isActive: PropTypes.bool,
  isActiveOnMount: PropTypes.bool,
  type: PropTypes.oneOf([
    'fade',
    'shiftInDown',
    'shiftInLeft',
    'shiftInUp',
    'slideDown',
    'slideRight',
    'slowFade',
    'zoom',
  ]),
};

Transition.defaultProps = {
  children: undefined,
  hasCSSTransitionMountOnEnter: undefined,
  hasCSSTransitionUnmountOnExit: undefined,
  isActive: false,
  isActiveOnMount: false,
  type: 'fade',
};

export { Transition };
