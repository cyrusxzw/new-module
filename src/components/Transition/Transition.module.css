@import 'styles/variables/index';

%transition-opacity-200ms-linear {
  transition: opacity 200ms linear;
}

%transition-opacity-1500ms-linear {
  transition: opacity 1500ms linear;
}

%transition-transform-500ms-ease-out-opacity-500ms-linear {
  transition: transform 500ms var(--easing-ease-out-cubic), opacity 500ms linear;
}

%transition-transform-1500ms-ease-out-opacity-1500ms-linear {
  transition: transform 1500ms var(--easing-ease-out-cubic),
    opacity 1500ms linear;
}

%transition-transform-800ms-ease-out-visibility-800ms-ease-out {
  transition: transform 800ms var(--easing-ease-out-cubic),
    visibility 800ms var(--easing-ease-out-cubic);
}

%transition-transform-400ms-ease-out-visibility-400ms-ease-out {
  transition: transform 400ms var(--easing-ease-out-cubic),
    visibility 400ms var(--easing-ease-out-cubic);
}

%transition-transform-400ms-ease-out-visibility-400ms-ease-decelerate {
  transition: transform 400ms var(--easing-ease-out-decelerate),
    visibility 400ms var(--easing-ease-out-decelerate);
}

%transition-transform-400ms-ease-out-visibility-400ms-ease-acelerate {
  transition: transform 350ms var(--easing-ease-out-acelerate),
    visibility 350ms var(--easing-ease-out-decelerate);
}

.fadeEnter {
  opacity: 0;
}

.fadeEnterActive {
  @extend %transition-opacity-200ms-linear;

  opacity: 1;
}

.fadeEnterDone {
  opacity: 1;
}

.fadeExit {
  opacity: 1;
}

.fadeExitActive {
  @extend %transition-opacity-200ms-linear;

  opacity: 0;
}

.fadeExitDone {
  opacity: 0;
}

.fadeInEnter {
  position: absolute !important;
  right: 0;
  left: 0;
  opacity: 0;
}

.fadeInEnterActive {
  @extend %transition-opacity-200ms-linear;

  position: relative;
  opacity: 1;
}

.fadeInEnterDone {
  opacity: 1;
}

.fadeInExit {
  opacity: 1;
}

.fadeInExitActive {
  @extend %transition-opacity-200ms-linear;

  opacity: 0;
}

.fadeInExitDone {
  position: absolute !important;
  right: 0;
  left: 0;
  opacity: 0;
}

.fixedEnter {
  position: relative;
}

.fixedEnterActive {
  position: fixed !important;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.fixedEnterDone {
  position: fixed;
}

.fixedExit {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.fixedExitActive {
  position: fixed !important;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.fixedExitDone {
  position: relative;
}

.shiftInDownEnter {
  opacity: 0;
  transform: translate3d(0, -20px, 0);
}

.shiftInDownEnterActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInDownEnterDone {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInDownExit {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInDownExitActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 0;
  transform: translate3d(0, -20px, 0);
}

.shiftInDownExitDone {
  opacity: 0;
  transform: translate3d(0, -20px, 0);
}

.shiftInUpEnter {
  opacity: 0;
  transform: translate3d(0, 20px, 0);
}

.shiftInUpEnterActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInUpEnterDone {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInUpExit {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInUpExitActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 0;
  transform: translate3d(0, 20px, 0);
}

.shiftInUpExitDone {
  opacity: 0;
  transform: translate3d(0, 20px, 0);
}

.shiftInLeftEnter {
  opacity: 0;
  transform: translate3d(20px, 0, 0);
}

.shiftInLeftEnterActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInLeftEnterDone {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInLeftExit {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInLeftExitActive {
  @extend %transition-transform-1500ms-ease-out-opacity-1500ms-linear;

  opacity: 0;
  transform: translate3d(20px, 0, 0);
}

.shiftInLeftExitDone {
  opacity: 0;
  transform: translate3d(20px, 0, 0);
}

.slideDownEnter {
  transform: translate3d(0, -100%, 0);
  visibility: hidden;
}

.slideDownEnterActive {
  @extend %transition-transform-800ms-ease-out-visibility-800ms-ease-out;

  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideDownEnterDone {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideDownExit {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideDownExitActive {
  @extend %transition-transform-800ms-ease-out-visibility-800ms-ease-out;

  transform: translate3d(0, -100%, 0);
  visibility: hidden;
}

.slideDownExitDone {
  transform: translate3d(0, -100%, 0);
  visibility: hidden;
}

.slideRightEnter {
  transform: translate3d(-100%, 0, 0);
  visibility: hidden;
}

.slideRightEnterActive {
  @extend %transition-transform-800ms-ease-out-visibility-800ms-ease-out;

  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideRightEnterDone {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideRightExit {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideRightExitActive {
  @extend %transition-transform-800ms-ease-out-visibility-800ms-ease-out;

  transform: translate3d(-100%, 0, 0);
  visibility: hidden;
}

.slideRightExitDone {
  transform: translate3d(-100%, 0, 0);
  visibility: hidden;
}

.slideLeftEnter {
  transform: translate3d(100%, 0, 0);
  visibility: hidden;
}

.slideLeftEnterActive {
  @extend %transition-transform-400ms-ease-out-visibility-400ms-ease-decelerate;

  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideLeftEnterDone {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideLeftExit {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideLeftExitActive {
  @extend %transition-transform-400ms-ease-out-visibility-400ms-ease-acelerate;

  transform: translate3d(100%, 0, 0);
  visibility: hidden;
}

.slideLeftExitDone {
  transform: translate3d(100%, 0, 0);
  visibility: hidden;
}

.slowFadeEnter {
  opacity: 0;
}

.slowFadeEnterActive {
  @extend %transition-opacity-1500ms-linear;

  opacity: 1;
}

.slowFadeEnterDone {
  opacity: 1;
}

.slowFadeExit {
  opacity: 1;
}

.slowFadeExitActive {
  @extend %transition-opacity-1500ms-linear;

  opacity: 0;
}

.slowFadeExitDone {
  opacity: 0;
}

.zoomEnter {
  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
}

.zoomEnterActive {
  @extend %transition-transform-500ms-ease-out-opacity-500ms-linear;

  opacity: 1;
  transform: scale3d(1, 1, 1);
}

.zoomEnterDone {
  opacity: 1;
  transform: scale3d(1, 1, 1);
}

.zoomExit {
  opacity: 1;
  transform: scale3d(1, 1, 1);
}

.zoomExitActive {
  @extend %transition-transform-500ms-ease-out-opacity-500ms-linear;

  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
}

.zoomExitDone {
  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
}
