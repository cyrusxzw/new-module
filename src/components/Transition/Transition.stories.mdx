import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import { Button } from '~/components/Button';
import { DefinitionList } from '~/components/DefinitionList';
import { DefinitionListFixture } from '~/components/DefinitionList/DefinitionList.fixture';
import { Transition } from './Transition';

<Meta title="Components/Transition" component={Transition} />

# Transition

<Canvas>
  <Story name="Base component">
    {(() => {
      const type = knobs.select(
        'type',
        [
          'fade',
          'shiftInDown',
          'shiftInLeft',
          'shiftInUp',
          'slideDown',
          'slideRight',
          'slideLeft',
          'slowFade',
          'zoom',
        ],
        'shiftInDown',
        'Presentation',
      );
      return (
        <State initialState={{ isActive: false }}>
          {(state, setState) => {
            const handleOnClick = () => {
              setState((state) => ({
                ...state,
                isActive: !state.isActive,
              }));
            };
            return (
              <React.Fragment>
                <Button onClick={handleOnClick} title="Toggle Transition">
                  Toggle Transition
                </Button>
                <Transition isActive={state.isActive} type={type}>
                  <DefinitionList items={DefinitionListFixture.items} />
                </Transition>
              </React.Fragment>
            );
          }}
        </State>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={Transition} />
