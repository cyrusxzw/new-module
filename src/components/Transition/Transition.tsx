import React, { isValidElement, ReactElement } from 'react';
import cx from 'classnames';
import get from 'lodash/get';
import { CSSTransition } from 'react-transition-group-npm';
import { useHasMounted } from '~/customHooks';
import { data } from './Transition.data';
import type { TransitionProps } from './Transition.types';

const Transition = ({
  children,
  isActive = false,
  isActiveOnMount = false,
  shouldMountOnEnter,
  shouldUnmountOnExit,
  type = 'fade',
}: TransitionProps): ReactElement | null => {
  const hasMounted = useHasMounted();

  if (!isValidElement(children)) {
    return null;
  }

  const isVisible = isActiveOnMount ? !!hasMounted : !!isActive;
  const classSet = cx(
    get(children, 'props.className', ''),
    isVisible ? '' : get(data[type], 'classNames.enter', ''),
  );

  return (
    <CSSTransition
      classNames={get(data[type], 'classNames', '')}
      in={isVisible}
      mountOnEnter={!!shouldMountOnEnter}
      timeout={get(data[type], 'timeout', 300)}
      unmountOnExit={!!shouldUnmountOnExit}
    >
      {React.cloneElement(children as ReactElement, {
        className: classSet,
      })}
    </CSSTransition>
  );
};

export { Transition };
