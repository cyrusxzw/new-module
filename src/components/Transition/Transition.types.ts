import type { ReactElement } from 'react';

type TransitionTypeLabel =
  | 'fade'
  | 'fadeIn'
  | 'shiftInDown'
  | 'shiftInLeft'
  | 'shiftInUp'
  | 'slideDown'
  | 'slideLeft'
  | 'slideRight'
  | 'slowFade'
  | 'fixed'
  | 'zoom';

type TransitionProps = {
  children: JSX.Element;
  isActive?: boolean;
  isActiveOnMount?: boolean;
  shouldMountOnEnter?: boolean;
  shouldUnmountOnExit?: boolean;
  type?: TransitionTypeLabel;
};

type TransitionType = (props: TransitionProps) => ReactElement | null;

export type { TransitionProps, TransitionType, TransitionTypeLabel };
