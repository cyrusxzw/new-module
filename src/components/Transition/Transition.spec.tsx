import React from 'react';
import { render } from '@testing-library/react';
import { Transition } from './Transition';

describe('<Transition />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <Transition isActive={true} type="fade">
        <div>Content</div>
      </Transition>,
    );

    expect(container).toMatchSnapshot();
  });
});
