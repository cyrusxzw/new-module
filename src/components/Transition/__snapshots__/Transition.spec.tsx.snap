// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transition /> should render base component correctly 1`] = `
<div>
  <div
    class=""
  >
    Content
  </div>
</div>
`;
